include/string.h

```diff
@@ -52,8 +56,9 @@ void *memchr (const void *src : itype(_Array_ptr<const void>) byte_count(n),
   : itype(_Array_ptr<void>) byte_count(n);
 char *strcpy (char *__restrict, const char *__restrict);
 // Copies the first n characters of source to destination. If the end of the source C string (which is signaled by a null-character) is found before num characters have been copied, destination is padded with zeros until a total of num characters have been written to it.
-char *strncpy(char *__restrict d : itype(__restrict _Nt_array_ptr<char>) count(n),
-              const char *__restrict s,
+char *strncpy(char *restrict d : itype(restrict _Nt_array_ptr<char>)
+                                 bounds((_Nt_array_ptr<char>)d, (_Nt_array_ptr<char>)d + n),
```

Comment:
AFAIK, `count(n)` should be equivalent to `bounds(d, d+n)`. Could you please add comments on why `count(n)` does not work here?

Reply:
When the above parameter d, whose type is qualified by __restrict, has its bounds declared as "count(n)" and is assigned to another pointer variable (whose type is not qualified by __restrict), the compiler complains that it cannot validate the bounds after initialization. Please see below for the outputs of the compiler on three tiny experimental code snippets.
The first two snippets illustrate subtle issues, and the third snippet compiles without warnings.
=====================================================================
sulekha@sulekha-VM:~/Work/checkedc-musl/build$ **cat foo.c**
void foo (void *__restrict p : itype(__restrict _Array_ptr<void>) byte_count(n), unsigned long n)
_Checked
{
        _Array_ptr<unsigned char> d : bounds((_Array_ptr<unsigned char>)p, (_Array_ptr<unsigned char>)p + n) = p;
        // ... more code
}
sulekha@sulekha-VM:~/Work/checkedc-musl/build$ **clang -c foo.c**
foo.c:4:35: warning: cannot prove declared bounds for 'd' are valid after initialization [-Wcheck-bounds-decls-checked-scope]
        _Array_ptr<unsigned char> d : bounds((_Array_ptr<unsigned char>)p,
        ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
foo.c:4:35: note: (expanded) declared bounds are 'bounds((_Array_ptr<unsigned char>)p, (_Array_ptr<unsigned char>)p + n)'
        _Array_ptr<unsigned char> d : bounds((_Array_ptr<unsigned char>)p,
                                  ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
foo.c:4:35: note: (expanded) inferred bounds are 'bounds((_Array_ptr<char>)p, (_Array_ptr<char>)p + n)'
1 warning generated.
NOTE: Even when I change the type of d in foo.c to _Array_ptr<char>, the compiler issues a similar warning.


=======================================================================
sulekha@sulekha-VM:~/Work/checkedc-musl/build$ **cat foo1.c**
void foo (void *__restrict p : itype(__restrict _Array_ptr<void>)
                                             bounds((_Array_ptr<void>)p, (_Array_ptr<void>)p + n),
               unsigned long n)
_Checked
{
        _Array_ptr<unsigned char> d : bounds((_Array_ptr<unsigned char>)p, (_Array_ptr<unsigned char>)p + n) = p;
        // ... more code
}
sulekha@sulekha-VM:~/Work/checkedc-musl/build$ **clang -c foo1.c**
foo1.c:3:59: error: arithmetic on a pointer to void
                                      (_Array_ptr<void>)p + n),
                                      ~~~~~~~~~~~~~~~~~~~ ^
1 error generated.


=========================================================================
sulekha@sulekha-VM:~/Work/checkedc-musl/build$ **cat foo2.c**
void foo (void *__restrict p : itype(__restrict _Array_ptr<void>)
                                             bounds((_Array_ptr<unsigned char>)p, (_Array_ptr<unsigned char>)p + n),
               unsigned long n)
_Checked
{
        _Array_ptr<unsigned char> d : bounds((_Array_ptr<unsigned char>)p, (_Array_ptr<unsigned char>)p + n) = p;
        // ... more code
}
sulekha@sulekha-VM:~/Work/checkedc-musl/build$ **clang -c foo2.c**